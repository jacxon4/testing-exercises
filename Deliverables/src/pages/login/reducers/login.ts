import {
  LoginEntity,
  LoginFormErrors,
  createEmptyLoginEntity,
  createEmptyLoginFormErrors
} from '../viewModel';
import { actionIds } from '../actions/actionIds';
export interface LoginState {
  loginEntity: LoginEntity;
  loginFormErrors: LoginFormErrors;
}

const loginStateFactory = (): LoginState => ({
  loginEntity: createEmptyLoginEntity(),
  loginFormErrors: createEmptyLoginFormErrors(),
});

export const loginReducer = (
  state: LoginState = loginStateFactory(),
  action
) => {
  switch (action.type) {
    case actionIds.UPDATE_LOGIN_ENTITY_FIELD:
      return handleUpdateLoginEntityField(state, action.payload);

    case actionIds.UPDATE_LOGIN_FORM_ERRORS:
      return handleUpdateloginFormErrors(state, action.payload);
  }
  return state;
};

const handleUpdateLoginEntityField = (
  state: LoginState,
  { fieldName, value, fieldValidationResult }
): LoginState => ({
  loginEntity: {
    ...state.loginEntity,
    [fieldName]: value,
  },
  loginFormErrors: {
    ...state.loginFormErrors,
    [fieldName]: fieldValidationResult,
  },
});

const handleUpdateloginFormErrors = (
  state: LoginState,
  { fieldErrors }
): LoginState => ({
  ...state,
  loginFormErrors: {
    ...state.loginFormErrors,
    ...fieldErrors,
  },
});
