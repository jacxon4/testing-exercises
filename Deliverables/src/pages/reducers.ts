import { combineReducers } from 'redux';
import { LoginState, loginReducer } from './login/reducers';
import { MembersState, membersReducer } from './members/list';
export interface State {
  login: LoginState;
  members: MembersState;
}

export const reducers = combineReducers<State>({
  login: loginReducer,
  members: membersReducer,
});
