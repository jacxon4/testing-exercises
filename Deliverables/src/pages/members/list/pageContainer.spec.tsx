import * as React from 'react';
import { shallow, mount } from 'enzyme';
import { MemberListPageContainer } from './index';
import configureStore from 'redux-mock-store';
import { State } from '../../reducers';
import * as fetchActions from './actions/fetchMembers';

const getMockStore = configureStore();

describe('pages/members/list/pageContainer', () => {
  let state;
  beforeAll(() => {
    state = {
      members: [],
    } as State;
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should render as expected when state member is empty', () => {
    const store = getMockStore(state);

    const component = shallow(<MemberListPageContainer />, {
      context: { store },
    });
    expect(component).toMatchSnapshot();
  });

  it('should provide the member list to the page when members exist', () => {
    state.members = [
      {
        id: 1,
        name: 'name',
        avatarUrl: 'avatar',
      },
    ];
    const store = getMockStore(state);
    //const stub = jest.spyOn(fetchActions, 'fetchMembers').mockReturnValue(members);

    const component = shallow(<MemberListPageContainer />, {
      context: { store },
    });
    expect(component).toMatchSnapshot();
  });

  it('should call the fech when page is mounted', async () => {
    const members = [
      {
        id: 1,
        name: 'name',
        avatarUrl: 'avatar',
      },
    ];
    state.members = members;
    const store = getMockStore(state);
    const stub = jest.spyOn(fetchActions, 'fetchMembers').mockReturnValue({
      type: 'test action type',
    });

    const component = mount(<MemberListPageContainer />, {
      context: { store },
    });

    await expect(stub).toHaveBeenCalled();
  });
});
