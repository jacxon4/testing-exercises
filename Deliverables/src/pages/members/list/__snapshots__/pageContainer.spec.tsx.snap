// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/members/list/pageContainer should provide the member list to the page when members exist 1`] = `
<PageContainer
  fetchMembers={[Function]}
  members={
    Array [
      Object {
        "avatarUrl": undefined,
        "id": 1,
        "name": undefined,
      },
    ]
  }
/>
`;

exports[`pages/members/list/pageContainer should render as expected when state member is empty 1`] = `
<PageContainer
  fetchMembers={[Function]}
  members={Array []}
/>
`;
